packages/core/src/modules/cache/singleContextLruCache/SingleContextStorageLruCache.ts

```diff
@@ -0,0 +1,161 @@
+
+    // Expired
+    if (item.expiresAt && Date.now() > item.expiresAt) {
+      cache.delete(key)
+      await this.persistCache(agentContext)
+      return null
+    }
+
+    return item.value as CacheValue
+  }
+
+  public async set<CacheValue>(
+    agentContext: AgentContext,
+    key: string,
+    value: CacheValue,
+    expiresInSeconds?: number
+  ): Promise<void> {
+    this.assertContextCorrelationId(agentContext)
+
+    let expiresDate = undefined
+
+    if (expiresInSeconds) {
+      expiresDate = new Date()
+      expiresDate.setSeconds(expiresDate.getSeconds() + expiresInSeconds)
+    }
+
+    const cache = await this.getCache(agentContext)
+    this.removeExpiredItems(cache)
+
+    cache.set(key, {
+      expiresAt: expiresDate?.getTime(),
+      value,
+    })
+    await this.persistCache(agentContext)
+  }
+
+  public async remove(agentContext: AgentContext, key: string): Promise<void> {
+    this.assertContextCorrelationId(agentContext)
+
+    const cache = await this.getCache(agentContext)
+    this.removeExpiredItems(cache)
+    cache.delete(key)
+
+    await this.persistCache(agentContext)
+  }
+
+  private async getCache(agentContext: AgentContext) {
+    if (!this._cache) {
+      const cacheRecord = await this.fetchCacheRecord(agentContext)
+      this._cache = this.lruFromRecord(cacheRecord)
+    }
+
+    return this._cache
+  }
+
+  private lruFromRecord(cacheRecord: CacheRecord) {
+    return new LRUMap<string, CacheItem>(
+      this.limit,
+      cacheRecord.entries.map((e) => [e.key, e.item])
+    )
+  }
+
+  private async fetchCacheRecord(agentContext: AgentContext) {
+    const cacheRepository = agentContext.dependencyManager.resolve(CacheRepository)
+    let cacheRecord = await cacheRepository.findById(agentContext, CONTEXT_STORAGE_LRU_CACHE_ID)
+
+    if (!cacheRecord) {
+      cacheRecord = new CacheRecord({
+        id: CONTEXT_STORAGE_LRU_CACHE_ID,
+        entries: [],
+      })
+
+      await cacheRepository.save(agentContext, cacheRecord)
+    }
+
+    return cacheRecord
+  }
+
+  private removeExpiredItems(cache: LRUMap<string, CacheItem>) {
+    cache.forEach((value, key) => {
+      if (value.expiresAt && Date.now() > value.expiresAt) {
+        cache.delete(key)
+      }
+    })
+  }
+
+  private async persistCache(agentContext: AgentContext) {
+    const cacheRepository = agentContext.dependencyManager.resolve(CacheRepository)
+    const cache = await this.getCache(agentContext)
+
+    await cacheRepository.update(
+      agentContext,
+      new CacheRecord({
+        entries: cache.toJSON().map(({ key, value: item }) => ({ key, item })),
+        id: CONTEXT_STORAGE_LRU_CACHE_ID,
+      })
+    )
+  }
+
+  /**
+   * Asserts this class is not used with multiple agent context instances.
+   */
+  private assertContextCorrelationId(agentContext: AgentContext) {
+    if (!this._contextCorrelationId) {
+      this._contextCorrelationId = agentContext.contextCorrelationId
+    }
```

Comment:
Would it make sense for the assert to also set the value? Feels a bit hacky.

Reply:
My reasoning was that it asserts it over time, and thus it keeps internal state to make sure it can make the assertion